akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Multiplexer.scala

```diff
@@ -16,116 +19,324 @@ import akka.util.ByteString
 
         currentInitialWindow = newDefaultWindow
-        outStreams.values.foreach(_.outboundWindowLeft += delta)
+        outStreams.values.foreach(i ⇒ updateWindowFor(i.streamId, delta))
       }
+      override def updatePriority(info: PriorityFrame): Unit = prioritizer.updatePriority(info)
 
       private def streamFor(streamId: Int): OutStream = outStreams.get(streamId) match {
         case None ⇒
-          val newOne = new OutStream(None, currentInitialWindow)
+          val newOne = new OutStream(streamId, None, currentInitialWindow)
           outStreams += streamId → newOne
           newOne
         case Some(old) ⇒ old
       }
       private def windowLeftFor(streamId: Int): Int = streamFor(streamId).outboundWindowLeft
-      private def updateWindowFor(streamId: Int, increment: Int): Unit = streamFor(streamId).outboundWindowLeft += increment
+      private def updateWindowFor(streamId: Int, increment: Int): Unit = {
+        val info = streamFor(streamId)
+        info.outboundWindowLeft += increment
+        if (info.canSend) enqueueOutStream(info)
+      }
 
-      override def doPush(elem: ElementAndTrigger): Unit =
-        elem.element match {
-          case d @ DataFrame(streamId, _, payload) ⇒
-            val sendSize = totalOutboundWindowLeft min windowLeftFor(streamId) min payload.size
+      def enqueueOutStream(outStream: OutStream): Unit = state.enqueueOutStream(outStream)
 
-            def send(elem: ElementAndTrigger): Unit = {
-              super.doPush(elem)
-              totalOutboundWindowLeft -= elem.element.asInstanceOf[DataFrame].payload.size
-              updateWindowFor(streamId, -elem.element.asInstanceOf[DataFrame].payload.size)
-            }
+      override def onDownstreamFinish(): Unit = {
+        outStreams.values.foreach(_.cancelStream())
+        completeStage()
+      }
+
+      var state: MultiplexerState = Idle
+      def onPull(): Unit = state.onPull()
+      private def become(nextState: MultiplexerState): Unit = {
+        if (nextState.name != state.name) recordStateChange(state.name, nextState.name)
+
+        state = nextState
+      }
+
+      sealed trait MultiplexerState extends Product {
+        def name: String = productPrefix
+
+        def onPull(): Unit
+        def pushControlFrame(frame: FrameEvent): Unit
+        def connectionWindowAvailable(): Unit
+        def enqueueOutStream(outStream: OutStream): Unit
+      }
+
+      // Multiplexer state machine
+      // Idle: No data to send, no demand from the network
+      // WaitingForData: Got demand from the network but no data to send
+      // WaitingForNetworkToSendControlFrames: Control frames (and maybe data frames) are queued but there is no network demand
+      // WaitingForNetworkToSendData: Data frames queued but no network demand
+      // WaitingForConnectionWindow: Data frames queued, demand from the network, but no connection-level window available
```

Comment:
state machine states look good; "network demand" we mean as RS demand, not enough-window -to-send-stuff I assume, since control may be sent without window impacat AFAIR; just sanity checking :)